File: src/python/optimized/attack.py
from numba import njit
from time import process_time

from src.optimized.key_schedule import key_schedule

K = 1 << 24


@njit
def encrypt(roundKeys, plain):
    """ PRESENT24 encryption of a plaintext with a 24-bit key.

    :param master_key: 24-bit key.
    :param plain: 24-bit plaintext.
    :return: cypher: PRESENT24 24-bit cypher.
    """
    SBOX = [0xc, 0x5, 0x6, 0xb,
            0x9, 0x0, 0xa, 0xd,
            0x3, 0xe, 0xf, 0x8,
            0x4, 0x7, 0x1, 0x2]
    cypher = plain
    for key in range(10):
        cypher = cypher ^ roundKeys[key]
        # Substitution
        tmp = 0x0
        for bitBloc in range(6):
            tmp |= SBOX[(cypher >> bitBloc * 4) & 0xf] << bitBloc * 4
        cypher = tmp
        # Permutation
        tmp = 0x0
        for bit in range(23):
            tmp |= ((cypher >> bit) & 0x1) << ((6 * bit) % 23)
        cypher = tmp | cypher & 0x800000

    cypher = cypher ^ roundKeys[-1]
    return cypher


@njit
def double_present24(key1, key2, plain):
    """ 2-PRESENT24 encryption of a plaintext with a 48-bit key (k1, k2). """
    return encrypt(key_schedule(key2), encrypt(key_schedule(key1), plain))


@njit
def decrypt(roundKeys, cypher):
    """ PRESENT24 decryption of a cypher with a 24-bit key.

    :param master_key: 24-bit key.
    :param cypher: PRESENT24 24-bit cypher.
    :return: plain: 24-bit decrypted plaintext.
    """
    SBOX = [0x5, 0xe, 0xf, 0x8,
            0xc, 0x1, 0x2, 0xd,
            0xb, 0x4, 0x6, 0x3,
            0x0, 0x7, 0x9, 0xa]
    plain = cypher ^ roundKeys[-1]

    for key in range(9, -1, -1):
        # Permutation
        tmp = 0x0
        for bit in range(23):
            tmp |= ((plain >> bit) & 0x1) << ((4 * bit) % 23)
        plain = tmp | plain & 0x800000
        # Substitution
        tmp = 0x0
        for bitBloc in range(6):
            tmp |= SBOX[(plain >> bitBloc * 4) & 0xf] << bitBloc * 4
        plain = tmp ^ roundKeys[key]
    return plain


@njit
def generate_intermediate_states(plain, cypher):
    """ Generate intermediate encryption and decryption lists for the MitM attack. """
    plainIntermediate, cypherIntermediate = [(0, 0)] * K, [(0, 0)] * K
    for masterKey in range(K):
        roundKeys = key_schedule(masterKey)
        plainIntermediate[masterKey] = (encrypt(roundKeys, plain), masterKey)
        cypherIntermediate[masterKey] = (decrypt(roundKeys, cypher), masterKey)
    return plainIntermediate, cypherIntermediate


@njit
def search_candidates(plainIntermediates, cypherIntermediates, plain, cypher):
    """ Find matching keys between two sorted lists of tuples.

    :return: candidateKeys: list of tuples of candidate keys (key1, key2).
    """
    plainCount, cypherCount = len(plainIntermediates), len(cypherIntermediates)
    i, j = 0, 0
    candidateKeys = []
    while i < plainCount and j < cypherCount:
        encrypted, decrypted = plainIntermediates[i][0], cypherIntermediates[j][0]
        if encrypted < decrypted:
            i += 1
        elif encrypted > decrypted:
            j += 1
        else:
            testCypher = double_present24(plainIntermediates[i][1], cypherIntermediates[j][1], plain)
            if testCypher == cypher:
                candidateKeys.append((plainIntermediates[i][1], cypherIntermediates[j][1]))
            i += 1
            j += 1
    return candidateKeys


@njit
def mitm_attack_present24(plain1, cypher1, plain2, cypher2):
    """ Perform a meet-in-the-middle attack on the PRESENT24 cipher to find candidate keys.
    """
    plainIntermediate, cypherIntermediate = generate_intermediate_states(plain1, cypher1)
    plainIntermediate.sort()
    cypherIntermediate.sort()

    candidateKeys = search_candidates(plainIntermediate, cypherIntermediate, plain2, cypher2)
    output(plain1, cypher1, plain2, cypher2, candidateKeys)


@njit
def output(plain1, cypher1, plain2, cypher2, candidateKeys):
    print(f"Plain-cypher used for the MitM: ({int_hex(plain1)},",
          f"{int_hex(cypher1)}) and ({int_hex(plain2)}, {int_hex(cypher2)})")
    if not candidateKeys:
        print("No candidate keys found.")
    else:
        print("Candidate keys found:")
        for key1, key2 in candidateKeys:
            print(f"- ({int_hex(key1)}, {int_hex(key2)})")


@njit
def int_hex(num):
    hexChars = "0123456789abcdef"
    hexa = ""
    while num > 0:
        hexa = hexChars[num & 0b1111] + hexa
        num = num // 16
    return "0x" + hexa


if __name__ == '__main__':
    plain1, cypher1 = 0xd41330, 0x2f4a58
    plain2, cypher2 = 0x9d0af2, 0x57c9d6

    start = process_time()
    res = mitm_attack_present24(plain1, cypher1, plain2, cypher2)
    print("Durée totale:", process_time() - start)

File: src/rust/attack.rs
use rayon::iter::{ParallelIterator, IntoParallelIterator};
use rayon::prelude::ParallelSliceMut;

use crate::rust::key_schedule::key_schedule;


const SBOX: [u8; 16] = [0xc, 0x5, 0x6, 0xb, 0x9, 0x0, 0xa, 0xd, 0x3, 0xe, 0xf, 0x8, 0x4, 0x7, 0x1, 0x2];
const INV_SBOX: [u8; 16] = [0x5, 0xe, 0xf, 0x8, 0xc, 0x1, 0x2, 0xd, 0xb, 0x4, 0x6, 0x3, 0x0, 0x7, 0x9, 0xa];
const K: u32 = 1 << 24;


/// PRESENT24 encryption of a plaintext with a 24-bit key.
fn encrypt(round_keys: &[u32], plaintext: u32) -> u32 {
    let mut cipher = plaintext;
    for i in 0..10 {
        cipher ^= round_keys[i];
        // Substitution
        let mut temp: u32 = 0;
        for j in 0..6 {
            temp |= (SBOX[((cipher >> (j * 4)) & 0xf) as usize] as u32) << (j * 4);
        }
        cipher = temp;
        // Permutation
        temp = 0;
        for j in 0..23 {
            temp |= ((cipher >> j) & 1) << ((6 * j) % 23);
        }
        cipher = temp | (cipher & 0x800000);
    }
    cipher ^ round_keys[10]
}


/// PRESENT24 decryption of a cypher with a 24-bit key.
fn decrypt(round_keys: &[u32], ciphertext: u32) -> u32 {
    let mut plain = ciphertext ^ round_keys[10];
    for i in (0..10).rev() {
        // Inverse permutation
        let mut temp = 0;
        for j in 0..23 {
            temp |= ((plain >> j) & 1) << ((4 * j) % 23);
        }
        plain = temp | (plain & 0x800000);

        // Inverse substitution
        temp = 0;
        for j in 0..6 {
            temp |= (INV_SBOX[((plain >> (j * 4)) & 0xf) as usize] as u32) << (j * 4);
        }
        plain = temp ^ round_keys[i];
    }
    plain
}


/// 2-PRESENT24 encryption of a plaintext with a 48-bit key (k1, k2).
fn double_present24(key1: u32, key2: u32, plaintext: u32) -> u32 {
    let round_keys1 = key_schedule(key1);
    let round_keys2 = key_schedule(key2);
    encrypt(&round_keys2, encrypt(&round_keys1, plaintext))
}


/// Generate intermediate encryption and decryption lists for the MitM attack.
/// 
/// Even though for efvery master_key we calculate 2 times key_schedule(master_key) it is still faster.
/// We didn't use a better method to parallelize because it wouldn't be a fair comparison.
/// With the better parallelization we achieved 0.8s (instead of 4.0s).
fn generate_intermediate_states(plaintext: u32, ciphertext: u32) -> (Vec<(u32, u32)>, Vec<(u32, u32)>) {
    let mut plain_intermediate: Vec<_> = (0..K).into_par_iter()
        .map(|master_key| {
            let round_keys = key_schedule(master_key);
            (encrypt(&round_keys, plaintext), master_key)
        })
        .collect();

    let mut cipher_intermediate: Vec<_> = (0..K).into_par_iter()
        .map(|master_key| {
            let round_keys = key_schedule(master_key);
            (decrypt(&round_keys, ciphertext), master_key)
        })
        .collect();

    plain_intermediate.par_sort_unstable();
    cipher_intermediate.par_sort_unstable();
    (plain_intermediate, cipher_intermediate)
}


/// Find matching keys between two sorted lists of tuples.
fn search_candidates(
    plain_intermediate: &[(u32, u32)],
    cipher_intermediate: &[(u32, u32)],
    plaintext: u32,
    ciphertext: u32
) -> Vec<(u32, u32)> {
    let (plain_count, cipher_count) = (plain_intermediate.len(), cipher_intermediate.len());
    let mut candidate_keys = Vec::new();
    let mut i = 0;
    let mut j = 0;

    while i < plain_count && j < cipher_count {
        match plain_intermediate[i].0.cmp(&cipher_intermediate[j].0) {
            std::cmp::Ordering::Less => i += 1,
            std::cmp::Ordering::Greater => j += 1,
            std::cmp::Ordering::Equal => {

                let key1 = plain_intermediate[i].1;
                let key2 = cipher_intermediate[j].1;
                if double_present24(key1, key2, plaintext) == ciphertext {
                    candidate_keys.push((key1, key2));
                }
                i += 1;
                j += 1;
            }
        }
    }
    candidate_keys
}


/// Perform a meet-in-the-middle attack on the PRESENT24 cipher to find candidate keys.
pub fn mitm_attack_present24(plain1: u32, cipher1: u32, plain2: u32, cipher2: u32) {
    let (plain_intermediate, cipher_intermediate) = generate_intermediate_states(plain1, cipher1);

    let candidate_keys = search_candidates(&plain_intermediate, &cipher_intermediate, plain2, cipher2);
    output_results(plain1, cipher1, plain2, cipher2, &candidate_keys);
}


fn output_results(plain1: u32, cipher1: u32, plain2: u32, cipher2: u32, candidate_keys: &[(u32, u32)]) {
    println!(
        "Plain-cipher used for the MitM: ({:#08x}, {:#08x}) and ({:#08x}, {:#08x})",
        plain1, cipher1, plain2, cipher2
    );

    if candidate_keys.is_empty() {
        println!("No candidate keys found.");
    } else {
        println!("Candidate keys found:");
        for (key1, key2) in candidate_keys {
            println!("- ({:#08x}, {:#08x})", key1, key2);
        }
    }
}
